network/src/gossip/message_routing.rs

```diff
@@ -0,0 +1,335 @@
+		for i in other_leaves {
+			for j in &self.leaves {
+				if i == j {
+					let leaf_topics = self.leaf_topics.get(i)
+						.expect("leaf_topics are mutated only in update_leaves; \
+							we have an entry for each item in self.leaves; \
+							i is in self.leaves; qed");
+
+					if leaf_topics.contains(topic) {
+						return true;
+					}
+				}
+			}
+		}
+
+		false
+	}
+
+	/// Get topics of all message queues a peer is interested in - this is useful
+	/// when a peer has informed us of their new best leaves.
+	pub fn intersection_topics(&self, other_leaves: &LeavesVec) -> impl Iterator<Item=Hash> {
+		let deduplicated = other_leaves.iter()
+			.filter_map(|l| self.leaf_topics.get(l))
+			.flat_map(|topics| topics.iter().cloned())
+			.collect::<HashSet<_>>();
+
+		deduplicated.into_iter()
+	}
+
+	/// Iterate over all live message queues for which the data is marked as not locally known,
+	/// calling a closure with `(topic, root)`. The closure will return whether the queue data is
+	/// unknown.
+	///
+	/// This is called when we should send un-routed message queues that we are
+	/// newly aware of to peers - as in when we update our leaves.
+	pub fn sweep_unknown_queues(&mut self, mut check_known: impl FnMut(&Hash, &Hash) -> bool) {
+		for (topic, &mut (ref queue_root, ref mut known)) in self.expected_queues.iter_mut() {
+			if !*known {
+				*known = check_known(topic, queue_root)
+			}
+		}
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::tests::TestChainContext;
+	use crate::gossip::{Known, GossipParachainMessages};
+	use polkadot_primitives::parachain::Message as ParachainMessage;
+
+	fn hash(x: u8) -> Hash {
+		[x; 32].into()
+	}
+
+	fn message_queue(from: u8, to: u8) -> Option<[[u8; 2]; 1]> {
+		if from == to {
+			None
+		} else {
+			Some([[from, to]])
+		}
+	}
+
+	fn message_queue_root(from: u8, to: u8) -> Option<Hash> {
+		message_queue(from, to).map(
+			|q| polkadot_validation::message_queue_root(q.iter())
+		)
+	}
+
+	fn check_roots(view: &mut View, i: u8, max: u8) -> bool {
+		for j in 0..max {
+			if let Some(messages) = message_queue(i, j) {
+				let queue_root = message_queue_root(i, j).unwrap();
+				let messages = GossipParachainMessages {
+					queue_root,
+					messages: messages.iter().map(|m| ParachainMessage(m.to_vec())).collect(),
+				};
+
+				match view.validate_queue_and_note_known(&messages).0 {
+					GossipValidationResult::ProcessAndKeep(topic) => if topic != queue_topic(queue_root) {
+						return false
+					},
+					_ => return false,
+				}
+			}
+		}
+
+		true
+	}
+
+	#[test]
+	fn update_leaves_none_in_common() {
+		let mut ctx = TestChainContext::default();
+		let max = 5;
+
+		for i in 0..max {
+			ctx.known_map.insert(hash(i as u8), Known::Leaf);
+
+			let messages_out: Vec<_> = (0..max).filter_map(|j| message_queue_root(i, j)).collect();
+
+			if !messages_out.is_empty() {
+				ctx.ingress_roots.insert(hash(i as u8), messages_out);
```

Comment:
If this is called `ingress_roots`, why does it contain a mapping `{from} -> HashSet<root>`?

Reply:
good point - i should switch `i` and `j`.

Reply:
probably renaming `i` and `j` to `from` and `to` would also help.

Reply:
no, wait, these are supposed to be the message queues available at a leaf. `i` is a leaf hash in this context, not a parachain ID

Reply:
ok, i see, in any case `i` and `j` should be switched, shouldn't they? and in `check_roots` also?

Reply:
i switched them, but then I un-switched them because they shouldn't be switched.
it just happens that I set up the test in a way where the leaf with `hash(i)` has messages going out from a parachain with ID `i`. nothing more to it than that.